compiler/rustc_monomorphize/src/collectable_trait.rs

```diff
@@ -0,0 +1,69 @@
+use rustc_hir::LangItem;
+use rustc_middle::mir::mono::MonoItem;
+use rustc_middle::ty::subst::GenericArg;
+use rustc_middle::ty::subst::SubstsRef;
+use rustc_middle::ty::{self, Ty, TyCtxt};
+use rustc_span::source_map::Spanned;
+use rustc_span::DUMMY_SP;
+
+/// Recurses through the component types of `ty` looking for types which
+/// implement the `Collectable` trait so that each
+/// `Collectable::set_collectable` method can be added to the monomorphization
+/// collector's output.
+pub(crate) fn collect_mono<'a, 'tcx>(
+    tcx: TyCtxt<'tcx>,
+    ty: Ty<'tcx>,
+    substs: SubstsRef<'tcx>,
+    output: &'a mut Vec<Spanned<MonoItem<'tcx>>>,
```

Comment:
Can we get duplicates here? Do we need a `HashSet` or similar?

Reply:
I think this is dealt with later by the mono collector. Either way, we don't really have much choice because `output` is the value that gets passed around the HIR visitor in the mono collector so we've just bolted the `collectable_trait.rs` functionality on top of that.

Reply:
OK.